setup/steps/02_ensure_gateway_provider.py

```diff
@@ -0,0 +1,427 @@
+            return result.returncode
+        
+        return 0
+        
+    except Exception as e:
+        # Ensure we restore directory even on exception
+        os.chdir(original_cwd)
+        announce(f"❌ Error running llmd-infra-installer.sh: {e}")
+        return 1
+
+
+def check_istio_crds_version(kubectl_cmd: str, dry_run: bool, verbose: bool) -> bool:
+    """
+    Check if Istio CRDs have v1 API version for workload entries.
+    
+    Args:
+        kubectl_cmd: kubectl command to use
+        dry_run: If True, return placeholder result
+        verbose: If True, print detailed output
+    
+    Returns:
+        bool: True if v1 CRDs exist, False otherwise
+    """
+    if dry_run:
+        announce("---> would check Istio CRD versions")
+        return True  # Assume OK in dry run
+    
+    try:
+        # Equivalent to: kubectl get crd -o "custom-columns=NAME:.metadata.name,VERSIONS:spec.versions[*].name" | grep -E "workload.*istio.*v1,"
+        cmd = [
+            kubectl_cmd, "get", "crd",
+            "-o", "custom-columns=NAME:.metadata.name,VERSIONS:spec.versions[*].name"
+        ]
+        
+        result = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
+        
+        if result.returncode != 0:
+            if verbose:
+                announce(f"❌ Failed to get CRDs: {result.stderr}")
+            return False
+        
+        # Check for workload entries with istio and v1 API
+        pattern = r'workload.*istio.*v1,'
+        has_v1_crds = bool(re.search(pattern, result.stdout))
+        
+        if verbose:
+            status = "found" if has_v1_crds else "not found"
+            announce(f"---> Istio workload CRDs with v1 API: {status}")
+        
+        return has_v1_crds
+        
+    except subprocess.TimeoutExpired:
+        announce("❌ kubectl get crd command timed out")
+        return False
+    except Exception as e:
+        announce(f"❌ Error checking Istio CRDs: {e}")
+        return False
+
+
+def apply_istio_crds(kubectl_cmd: str, dry_run: bool, verbose: bool) -> int:
+    """
+    Apply Istio CRDs from GitHub if needed.
+    
+    Args:
+        kubectl_cmd: kubectl command to use
+        dry_run: If True, only print what would be executed
+        verbose: If True, print detailed output
+    
+    Returns:
+        int: 0 for success, non-zero for failure
+    """
+    crd_url = "https://raw.githubusercontent.com/istio/istio/refs/tags/1.23.1/manifests/charts/base/crds/crd-all.gen.yaml"
+    
+    # Use llmdbench_execute_cmd to handle retries and dry run consistently
+    cmd = f"{kubectl_cmd} apply -f {crd_url}"
+    
+    announce("📜 Applying more recent CRDs (v1.23.1) from istio...")
+    result = llmdbench_execute_cmd(
+        actual_cmd=cmd,
+        dry_run=dry_run,
+        verbose=verbose,
+        silent=not verbose,
+        retry_interval=0,  # No retry interval
+        max_retries=3      # 3 retries as in original
```

Comment:
`max_retries` is not a parameter to `llmdbench_execute_cmd`. Perhaps `attempts`.